server/schedulers/hot_region.go

```diff
@@ -521,12 +561,52 @@ func (bs *balanceSolver) solve() []*operator.Operator {
 				bs.cur.dstStore = dstStore
 				bs.calcProgressiveRank()
 				tryUpdateBestSolution(isUniformFirstPriority)
+
+				if searchRevertRegions && (bs.cur.progressiveRank >= -1 && bs.cur.progressiveRank <= 0) &&
+					(bs.best == nil || bs.best.progressiveRank >= -1 || len(bs.best.revertRegions) > 0) {
+					// The search-revert-regions is performed only when the following conditions are met to improve performance.
+					// * `searchRevertRegions` is true. It depends on the result of the last `solve`.
+					// * `IsStrictPickingStoreEnabled` is false.
+					// * The current solution is not good enough. -1 <= progressiveRank <= 0
+					// * The current best solution is not good enough.
+					//     * The current best solution has progressiveRank < -1 and does not contain revert regions.
+					//     * The current best solution contain revert regions.
+					schedulerCounter.WithLabelValues(bs.sche.GetName(), "search-revert-regions").Inc()
+					dstStoreID := dstStore.GetID()
+					for _, dstPeerStat := range bs.filterHotPeers(bs.cur.dstStore) {
+						revertRegion := bs.getRegion(dstPeerStat, dstStoreID)
+						if revertRegion == nil || revertRegion.GetID() == bs.cur.region.GetID() ||
+							!allowRevertRegion(revertRegion, srcStoreID) {
+							continue
+						}
+						bs.cur.dstPeersStat = []*statistics.HotPeerStat{dstPeerStat}
+						bs.cur.revertRegions = []*core.RegionInfo{revertRegion}
+						bs.calcProgressiveRank()
+						tryUpdateBestSolution(isUniformFirstPriority)
+					}
+					bs.cur.dstPeersStat = nil
+					bs.cur.revertRegions = nil
+				}
 			}
 		}
 	}
+	bs.logBestSolution()
+	searchRevertRegions = bs.allowSearchRevertRegions()
+	bs.sche.searchRevertRegions[bs.resourceTy] = searchRevertRegions
+	if searchRevertRegions {
+		schedulerCounter.WithLabelValues(bs.sche.GetName(), "allow-search-revert-regions").Inc()
+	}
 	return bs.ops
 }
 
+func (bs *balanceSolver) allowSearchRevertRegions() bool {
+	// The next solve is allowed to search-revert-regions only when the following conditions are met.
+	// * No best solution was found this time.
+	// * The progressiveRank of the best solution is -1.
+	// * The best solution contain revert regions.
+	return bs.best == nil || bs.best.progressiveRank >= -1 || len(bs.best.revertRegions) > 0
+}
+
```

Comment:
For the multi operator, pending should be the result of taking the absolute value.